### ISSUE TITLE
sql: table lease leakage in 1.0.6

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#20422]
Forked from https://github.com/cockroachdb/cockroach/issues/20114#issuecomment-348704450:

I recompiled 1.0.6 with LeaseDuration set to 10s and MinLeaseDuration set to 2s in order to reproduce the badness seen in #20114. Running `kv` against a 5-node cluster for a few minutes results in this oddity:

```
root@:26257/system> SELECT count(*) FROM system.lease;
+----------+
| count(*) |
+----------+
|       36 |
+----------+
```

The load I'm running is hitting a single table. There should be 1-2 table leases per node. So there should be 5-10 leases. It appears that old leases are not being cleaned up correctly. It is possible this is a 1.0.x specific bug.

I think I see where the leakage is coming from. If `tableState.acquire()` finds a lease which has expired, it will not return it and instead acquire a new lease. But there doesn't seem to be any code which will clean up the old lease if `LeaseState.refcount` is 0. Actually, there is a bit of code which does that clean up, but it is only run if the `SystemConfig` changes. See `LeaseManager.RefreshLeases`.

I just verified that creating a new table (which results in a new `SystemConfig`) results in all of the old leases being purged:

```
root@:26257/> create table test.t (t INT);
CREATE TABLE
root@:26257/> SELECT count(*) FROM system.lease;
+----------+
| count(*) |
+----------+
|        5 |
+----------+
(1 row)
```

Cc @a-robinson

**Comment by petermattis** [comment]
The leak isn't present on master. The table lease code has changed significantly. In particular, replacing a lease with a new one for the same table version now results on the old lease being released immediately (as the node now has 2 leases for the same table version, only 1 is necessary).

**Comment by petermattis** [comment]
Looks like 1.1 also doesn't contain the leak (I tested master, haven't tested 1.1). The code in 1.1 is similar to the code in master.

**Comment by petermattis** [comment]
Looks like https://github.com/cockroachdb/cockroach/pull/16453 was where this was fixed.

**Comment by a-robinson** [comment]
Thanks for doing this testing! Is the thought that we'd like to cherry-pick something back to 1.0? Or should this be closed given that 1.1 and head contain a fix?

**Comment by petermattis** [comment]
The change in 1.1 which fixed the table lease leak is large enough that I'd be worried about cherry-picking it. Let's wait for the user to speak up regarding whether the workaround (periodically setting a cluster setting) is sufficient.

**Comment by vivekmenezes** [comment]
Closing this issue because it's not in 1.1 and forward

**Comment by bdarnell** [comment]
Reopening this because we want to do something about it in 1.0.7. Maybe that's backporting #16453, or maybe it's a smaller more targetted fix.

**Comment by bdarnell** [comment]
Fixed in #22563
